File: Algorithms/p54_Spiral_Matrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        if(matrix.size()==0) return {};
        vector<int> res(matrix.size()*matrix[0].size());
        int l = 0;
        int r = matrix[0].size()-1;
        int u = 0;
        int d = matrix.size()-1;
        int i = 0;
        while(1){
            if(l>r) break;
            for(int k=l;k<=r;++k){
                res[i++] = matrix[u][k];
            }
            ++u;
            if(u>d) break;
            for(int k=u;k<=d;++k){
                res[i++] = matrix[k][r];
            }
            --r;
            if(l>r) break;
            for(int k=r;k>=l;--k){
                res[i++] = matrix[d][k];
            }
            --d;
            if(u>d) break;
            for(int k=d;k>=u;--k){
                res[i++] = matrix[k][l];
            }
            ++l;
        }
        return res;
    }
};

File: Algorithms/p54_Spiral_Matrix.py
class Solution(object):
    def spiralOrder(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[int]
        """
        if len(matrix)==0:
            return []
        res = [None]*(len(matrix)*len(matrix[0]))
        l = 0
        r = len(matrix[0])-1
        u = 0
        d = len(matrix)-1
        i = 0
        while True:
            if l>r:
                break
            for k in range(l,r+1):
                res[i] = matrix[u][k]
                i += 1
            u += 1
            if u>d:
                break
            for k in range(u,d+1):
                res[i] = matrix[k][r]
                i += 1
            r -= 1
            if l>r:
                break
            for k in range(r,l-1,-1):
                res[i] = matrix[d][k]
                i += 1
            d -= 1
            if u>d:
                break
            for k in range(d,u-1,-1):
                res[i] = matrix[k][l]
                i += 1
            l += 1
        return res
